scipy/stats/_stats_py.py

```diff
@@ -3911,28 +3911,46 @@ def trim_mean(a, proportiontocut, axis=0):
 F_onewayResult = namedtuple('F_onewayResult', ('statistic', 'pvalue'))
 
 
-def _create_f_oneway_nan_result(shape, axis):
+def _create_f_oneway_nan_result(shape, axis, samples):
     """
     This is a helper function for f_oneway for creating the return values
     in certain degenerate conditions.  It creates return values that are
     all nan with the appropriate shape for the given `shape` and `axis`.
     """
     axis = normalize_axis_index(axis, len(shape))
     shp = shape[:axis] + shape[axis+1:]
-    if shp == ():
-        f = np.nan
-        prob = np.nan
-    else:
-        f = np.full(shp, fill_value=np.nan)
-        prob = f.copy()
-    return F_onewayResult(f, prob)
+    f = np.full(shp, fill_value=_get_nan(*samples))
+    prob = f.copy()
+    return F_onewayResult(f[()], prob[()])
 
 
 def _first(arr, axis):
     """Return arr[..., 0:1, ...] where 0:1 is in the `axis` position."""
     return np.take_along_axis(arr, np.array(0, ndmin=arr.ndim), axis)
 
 
+def _f_oneway_is_too_small(samples, kwargs, axis=-1):
+    # Check this after forming alldata, so shape errors are detected
+    # and reported before checking for 0 length inputs.
+    if any(sample.shape[axis] == 0 for sample in samples):
+        msg = ('all input arrays have length 1.  f_oneway requires that at '
+               'least one input has length greater than 1.')
```

Comment:
```suggestion
        msg = 'at least one input has length 0'
```
Right?

Reply:
@tirthasheshpatel I'll go ahead and just commit this unresolved comment; I think it was just a copy-paste mistake.

Reply:
Oh, sorry, missed this one.

> Surprising that it's not causing a test failure, though.

Yeah, I noticed that some of the functions in stats don't have good coverage.

Reply:
Weird. In `test_stats.py` there is a test for this:
```python3
    def test_length0_1d_error(self):
        # Require at least one value in each group.
        msg = 'all input arrays have length 1.'
        with assert_warns(stats.DegenerateDataWarning, match=msg):
            result = stats.f_oneway([1, 2, 3], [], [4, 5, 6, 7])
            assert_equal(result, (np.nan, np.nan))
```
The test _shouldn't_ have been passing before this PR because it has the wrong message. I just corrected the warning message here without correcting the test, and the test still passes. So `assert_warns` seems to be ignoring the `match` parameter (or doing something that I don't understand).

Reply:
Yup, just checked with this MRE:

```pytest
import warnings
import numpy as np

def func():
  warnings.warn('1', UserWarning)

def test_func():
  with np.testing.assert_warns(UserWarning, match='0'):
    func()
```

Seems like a bug in NumPy.

Reply:
I opened gh-20031 on the SciPy side. Maybe I'll open a NumPy report later.